migration/README.md

```diff
@@ -1,17 +1,51 @@
-# Migration steps
+# Upgrade and migration
 
-This directory contains the migration steps for each major and minor version, including the documentation with instructions and the snippets for automation of the upgrade process.
+This directory contains the upgrade and migration steps for each major and minor version, including the documentation with instructions and the snippets for automation of the upgrade process.
 
 > [!note]
 > Only the five latest major and minor releases are kept, if you want to see the migration guides for older versions then you need to check out that version of the repository.
 
-## Migration template and script
+## Writing migration steps for contributions
 
-> [!NOTE]
-> Only major releases are supported by the script
+Notice to developers on writing migration steps:
 
-There is a script that can help you create the initial migration folder and `upgrade-apps.md` document. You can use the following command to run it from the root of compliantkubernetes-apps directory:
+- Migration steps:
+  - are written per minor version and placed in a subdirectory of the migration directory with the name `vX.Y/`,
+  - are written to be idempotent and usable no matter which patch version you are upgrading from and to,
+  - are documented in this document to be able to run them manually,
+  - are divided into prepare and apply steps:
+    - Prepare steps:
+      - are placed in the `prepare/` directory,
+      - may **only** modify the configuration of the environment,
+      - may **not** modify the state of the environment,
+      - steps are run in order of their names use two digit prefixes.
+    - Apply steps:
+      - are placed in the `apply/` directory,
+      - may **only** modify the state of the environment,
+      - may **not** modify the configuration of the environment,
+      - are run in order of their names use two digit prefixes,
+      - are run with the argument `execute` on upgrade and should return 1 on failure and 2 on successful internal rollback,
+      - are rerun with the argument `rollback` on execute failure and should return 1 on failure.
+
+For prepare the init step is given.
+For apply the bootstrap and the apply steps are given, it is expected that releases upgraded in custom steps are excluded from the apply step.
```

Comment:
Could maybe expand this a bit more and clarify how and where you are supposed to exclude certain steps, just to make it easier for people that are writing their first migration script :)

Reply:
This has been expanded on.